### ISSUE TITLE
access control exception if programmatic security manager is used

### ISSUE THREAD

**Issue by maggu2810** [eclipse/jetty.project#3653]
I have used a third party code that injects a security manager programmatically.
See:
https://github.com/openhab/openhab-core/blob/a37ccea/bundles/org.openhab.core.boot/src/main/java/org/openhab/core/internal/CoreActivator.java#L51-L59

After this part of third party code has been used the servlet registration failed:

```
2019-04-12T16:28:45,240 | ERROR | ESH-thingHandler-1 | HttpServiceStarted               | 327 - org.ops4j.pax.web.pax-web-runtime - 7.2.8 | Could not start the servlet context for context path []
java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "getClassLoader")
        at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472) ~[?:?]
        at java.security.AccessController.checkPermission(AccessController.java:884) ~[?:?]
        at org.eclipse.jetty.server.handler.ContextHandler$Context.getClassLoader(ContextHandler.java:2617) ~[?:?]
        at org.eclipse.jetty.websocket.jsr356.server.deploy.WebSocketServerContainerInitializer.onStartup(WebSocketServerContainerInitializer.java:210) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext$1.call(HttpServiceContext.java:218) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext$1.call(HttpServiceContext.java:213) ~[?:?]
        at org.ops4j.pax.swissbox.core.ContextClassLoaderUtils.doWithClassLoader(ContextClassLoaderUtils.java:60) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.lambda$doStart$1(HttpServiceContext.java:212) ~[?:?]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doStart(HttpServiceContext.java:210) ~[?:?]
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerImpl$1.start(JettyServerImpl.java:329) ~[?:?]
        at org.ops4j.pax.web.service.internal.HttpServiceStarted.registerServlet(HttpServiceStarted.java:252) ~[?:?]
        at org.ops4j.pax.web.service.internal.HttpServiceStarted.registerServlet(HttpServiceStarted.java:223) ~[?:?]
        at org.ops4j.pax.web.service.internal.HttpServiceStarted.registerServlet(HttpServiceStarted.java:207) ~[?:?]
        at org.ops4j.pax.web.service.internal.HttpServiceProxy.registerServlet(HttpServiceProxy.java:69) ~[?:?]
        at foo.bar.Something.open(Something.java:79) ~[?:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
2019-04-12T16:28:45,268 | ERROR | FelixStartLevel  | HttpServiceStarted               | 327 - org.ops4j.pax.web.pax-web-runtime - 7.2.8 | Could not start the servlet context for context path []
java.security.AccessControlException: access denied ("java.lang.RuntimePermission" "getClassLoader")
        at java.security.AccessControlContext.checkPermission(AccessControlContext.java:472) ~[?:?]
        at java.security.AccessController.checkPermission(AccessController.java:884) ~[?:?]
        at org.eclipse.jetty.server.handler.ContextHandler$Context.getClassLoader(ContextHandler.java:2617) ~[?:?]
        at org.eclipse.jetty.websocket.jsr356.server.deploy.WebSocketServerContainerInitializer.onStartup(WebSocketServerContainerInitializer.java:210) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext$1.call(HttpServiceContext.java:218) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext$1.call(HttpServiceContext.java:213) ~[?:?]
        at org.ops4j.pax.swissbox.core.ContextClassLoaderUtils.doWithClassLoader(ContextClassLoaderUtils.java:60) ~[327:org.ops4j.pax.web.pax-web-runtime:7.2.8]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.lambda$doStart$1(HttpServiceContext.java:212) ~[?:?]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doStart(HttpServiceContext.java:210) ~[?:?]
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:68) ~[?:?]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerImpl$1.start(JettyServerImpl.java:329) ~[?:?]
        at org.ops4j.pax.web.service.internal.HttpServiceStarted.registerServlet(HttpServiceStarted.java:252) [327:org.ops4j.pax.web.pax-web-runtime:7.2.8]
        at org.ops4j.pax.web.service.internal.HttpServiceStarted.registerServlet(HttpServiceStarted.java:223) [327:org.ops4j.pax.web.pax-web-runtime:7.2.8]
        at org.ops4j.pax.web.service.internal.HttpServiceStarted.registerServlet(HttpServiceStarted.java:207) [327:org.ops4j.pax.web.pax-web-runtime:7.2.8]
        at org.ops4j.pax.web.service.internal.HttpServiceProxy.registerServlet(HttpServiceProxy.java:69) [327:org.ops4j.pax.web.pax-web-runtime:7.2.8]
        at org.openhab.core.internal.icon.IconForwarder.setHttpService(IconForwarder.java:44) [269:org.openhab.core.boot:2.5.0.a3]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invokeMethod(BaseMethod.java:228) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.access$500(BaseMethod.java:41) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod$Resolved.invoke(BaseMethod.java:664) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.inject.methods.BaseMethod.invoke(BaseMethod.java:510) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.inject.methods.BindMethod.invoke(BindMethod.java:42) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.DependencyManager.doInvokeBindMethod(DependencyManager.java:1813) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.DependencyManager.bindDependency(DependencyManager.java:1637) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.DependencyManager.bind(DependencyManager.java:1624) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createImplementationObject(SingleComponentManager.java:307) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.createComponent(SingleComponentManager.java:114) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getService(SingleComponentManager.java:982) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.SingleComponentManager.getServiceInternal(SingleComponentManager.java:955) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.activateInternal(AbstractComponentManager.java:765) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.enableInternal(AbstractComponentManager.java:666) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.AbstractComponentManager.enable(AbstractComponentManager.java:432) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.manager.ConfigurableComponentHolder.enableComponents(ConfigurableComponentHolder.java:665) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.BundleComponentActivator.initialEnable(BundleComponentActivator.java:338) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.Activator.loadComponents(Activator.java:382) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.Activator.access$200(Activator.java:49) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.Activator$ScrExtension.start(Activator.java:264) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.AbstractExtender.createExtension(AbstractExtender.java:196) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:169) [133:org.apache.felix.scr:2.1.16]
        at org.apache.felix.scr.impl.AbstractExtender.modifiedBundle(AbstractExtender.java:49) [133:org.apache.felix.scr:2.1.16]
        at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:482) [?:?]
        at org.osgi.util.tracker.BundleTracker$Tracked.customizerModified(BundleTracker.java:415) [?:?]
        at org.osgi.util.tracker.AbstractTracked.track(AbstractTracked.java:232) [?:?]
        at org.osgi.util.tracker.BundleTracker$Tracked.bundleChanged(BundleTracker.java:444) [?:?]
        at org.apache.felix.framework.EventDispatcher$3.run(EventDispatcher.java:908) [?:?]
        at java.security.AccessController.doPrivileged(Native Method) ~[?:?]
        at org.apache.felix.framework.EventDispatcher.invokeBundleListenerCallback(EventDispatcher.java:904) [?:?]
        at org.apache.felix.framework.EventDispatcher.fireEventImmediately(EventDispatcher.java:834) [?:?]
        at org.apache.felix.framework.EventDispatcher.fireBundleEvent(EventDispatcher.java:516) [?:?]
        at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4579) [?:?]
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2174) [?:?]
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1373) [?:?]
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
```

While reading the Jetty code I realized that Jetty is using

```
AccessController.checkPermission(new RuntimePermission("getClassLoader"));
```

if a security manager is used (https://github.com/eclipse/jetty.project/blob/jetty-9.4.12.v20180830/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ContextHandler.java#L2617).

If my understanding of the security manager topic is correct `AccessController.checkPermission` is normally called by the default security manager if used.

Shouldn't -- to support also programmatic security manager -- the code look like:

```
SecurityManager security = System.getSecurityManager();
if (security != null) {
    security.checkPermission(new RuntimePermission("getClassLoader"));
}
```

Otherwise the `checkPermission` of the above registered security manager is not called at all.

So, is my understanding correct, that Jetty calls `AccessController.checkPermission` if a security manager is used instead of calling the `checkPermission` of the used security manager?

**Comment by janbartel** [comment]
@maggu2810 yes I think you're right. See PR #3654.

**Comment by janbartel** [comment]
Fix applied for jetty-9.4.19 and beyond.
